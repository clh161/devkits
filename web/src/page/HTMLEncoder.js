// @flow
import React, { useState } from 'react';
import type { Node } from 'react';
import { Divider, Grid, TextareaAutosize } from '@material-ui/core';
import { encode, decode } from 'html-entities';

type Props = {};
const DEFAULT_TEXT = 'Example: < > " \\\\\' &';

export default function HTMLEncoder(props: Props): Node {
  const [decodedHTMLText, setDecodedHTMLText] = useState<string>(DEFAULT_TEXT);
  const [encodedHTMLText, setEncodedHTMLText] = useState<string>(
    encode(DEFAULT_TEXT)
  );

  function onEncodedHTMLChanged(event) {
    const { value } = event.target;
    setEncodedHTMLText(value);
    setDecodedHTMLText(decode(value));
  }
  function onDecodedHTMLChanged(event) {
    const { value } = event.target;
    setDecodedHTMLText(value);
    setEncodedHTMLText(encode(value));
  }
  return (
    <Grid container spacing={3}>
      <Grid item xs={12}>
        <TextareaAutosize
          onChange={onDecodedHTMLChanged}
          style={{ width: '100%', minHeight: 160 }}
          value={decodedHTMLText}
          placeholder="Decoded HTML"
        />
      </Grid>
      <Divider />
      <Grid item xs={12}>
        <TextareaAutosize
          style={{ width: '100%', minHeight: 160 }}
          placeholder="Encoded HTML"
          value={encodedHTMLText}
          onChange={onEncodedHTMLChanged}
        />
      </Grid>
    </Grid>
  );
}
