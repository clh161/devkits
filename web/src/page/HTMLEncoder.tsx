import { Divider, Grid } from '@mui/material';
import { decode, encode } from 'html-entities';
import React, { ReactElement, useState } from 'react';
type Props = {
  initDecodedText?: string;
  initEncodedText?: string;
};
export default function HTMLEncoder({
  initDecodedText,
  initEncodedText,
}: Props): ReactElement {
  const [decodedHTMLText, setDecodedHTMLText] = useState<string>(
    initDecodedText ?? decode(initEncodedText)
  );
  const [encodedHTMLText, setEncodedHTMLText] = useState<string>(
    initEncodedText ?? encode(initDecodedText)
  );

  function onEncodedHTMLChanged(event) {
    const { value } = event.target;
    setEncodedHTMLText(value);
    setDecodedHTMLText(decode(value));
  }

  function onDecodedHTMLChanged(event) {
    const { value } = event.target;
    setDecodedHTMLText(value);
    setEncodedHTMLText(encode(value));
  }

  return (
    <Grid container spacing={3}>
      <Grid item xs={12}>
        <textarea
          onChange={onDecodedHTMLChanged}
          placeholder='Decoded HTML'
          style={{
            width: '100%',
            minHeight: 160,
          }}
          value={decodedHTMLText}
        />
      </Grid>
      <Divider />
      <Grid item xs={12}>
        <textarea
          onChange={onEncodedHTMLChanged}
          placeholder='Encoded HTML'
          style={{
            width: '100%',
            minHeight: 160,
          }}
          value={encodedHTMLText}
        />
      </Grid>
    </Grid>
  );
}
