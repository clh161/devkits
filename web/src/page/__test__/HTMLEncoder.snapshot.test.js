// @flow

import React from 'react';
import { Helmet } from 'react-helmet';
import renderer from 'react-test-renderer';

import HTMLEncoder from '../HTMLEncoder';

it('Init with decoded text', () => {
  const tree = renderer
    .create(<HTMLEncoder initDecodedText={'Example: < > " \\\' &'} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Init with encoded text', () => {
  const tree = renderer
    .create(
      <HTMLEncoder
        initEncodedText={'Example: &lt; &gt; &quot; \\&apos; &amp;'}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
  expect(Helmet.peek()).toMatchSnapshot();
});
