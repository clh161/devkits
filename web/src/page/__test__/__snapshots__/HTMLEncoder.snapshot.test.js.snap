// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Init with decoded text 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h1
        className="MuiTypography-root MuiTypography-h4"
      >
        HTML Encoder
      </h1>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <textarea
        onChange={[Function]}
        placeholder="Decoded HTML"
        style={
          Object {
            "minHeight": 160,
            "width": "100%",
          }
        }
        value="Example: < > \\" \\\\' &"
      />
    </div>
    <hr
      className="MuiDivider-root"
    />
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <textarea
        onChange={[Function]}
        placeholder="Encoded HTML"
        style={
          Object {
            "minHeight": 160,
            "width": "100%",
          }
        }
        value="Example: &lt; &gt; &quot; \\\\&apos; &amp;"
      />
    </div>
  </div>
</div>
`;

exports[`Init with encoded text 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h1
        className="MuiTypography-root MuiTypography-h4"
      >
        HTML Encoder
      </h1>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <textarea
        onChange={[Function]}
        placeholder="Decoded HTML"
        style={
          Object {
            "minHeight": 160,
            "width": "100%",
          }
        }
        value="Example: < > \\" \\\\' &"
      />
    </div>
    <hr
      className="MuiDivider-root"
    />
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <textarea
        onChange={[Function]}
        placeholder="Encoded HTML"
        style={
          Object {
            "minHeight": 160,
            "width": "100%",
          }
        }
        value="Example: &lt; &gt; &quot; \\\\&apos; &amp;"
      />
    </div>
  </div>
</div>
`;

exports[`Init with encoded text 2`] = `
Object {
  "baseTag": Array [],
  "bodyAttributes": Object {},
  "defer": true,
  "encode": true,
  "htmlAttributes": Object {},
  "linkTags": Array [],
  "metaTags": Array [
    Object {
      "content": "Encode or decode HTML into text.",
      "name": "description",
    },
    Object {
      "content": "HTML Encoder, HTML Decoder, encoding HTML, decoding HTML",
      "name": "keywords",
    },
  ],
  "noscriptTags": Array [],
  "onChangeClientState": [Function],
  "scriptTags": Array [],
  "styleTags": Array [],
  "title": "HTML Encoder",
  "titleAttributes": Object {},
}
`;
